vng/dict.go

```diff
@@ -0,0 +1,143 @@
+package vng
+
+import (
+	"io"
+	"math"
+
+	"github.com/brimdata/super"
+	"github.com/brimdata/super/zcode"
+	"golang.org/x/sync/errgroup"
+)
+
+type DictEncoder struct {
+	typ    super.Type
+	tags   map[string]byte
+	index  []byte
+	counts *Uint32Encoder
+	values Encoder
+}
+
+func NewDictEncoder(typ super.Type, values Encoder) *DictEncoder {
+	if _, ok := values.(resetter); !ok {
+		panic("Dict values encoder must be resettable")
+	}
+	var tags map[string]byte
+	if id := typ.ID(); id != super.IDUint8 && id != super.IDInt8 && id != super.IDBool {
+		// Don't bother using a dictionary (which takes 8-bit tags) to encode
+		// other 8-bit values.
+		tags = make(map[string]byte)
+	}
+	return &DictEncoder{
+		typ:    typ,
+		tags:   tags,
+		values: values,
+	}
+}
+
+func (d *DictEncoder) Write(body zcode.Bytes) {
+	d.values.Write(body)
+	if d.tags != nil {
+		tag, ok := d.tags[string(body)]
+		if !ok {
+			len := len(d.tags)
+			if len > math.MaxUint8 {
+				d.tags = nil
+				return
+			}
+			tag = byte(len)
+			d.tags[string(body)] = tag
+		}
+		d.index = append(d.index, tag)
+	}
+}
+
+func (d *DictEncoder) Const() *Const {
+	if len(d.tags) != 1 {
+		return nil
+	}
+	var bytes zcode.Bytes
+	for b := range d.tags {
+		bytes = []byte(b)
+	}
+	return &Const{
+		Value: super.NewValue(d.typ, bytes),
+		Count: uint32(len(d.index)),
+	}
+}
+
+func (d *DictEncoder) isValid() bool {
+	return d.tags != nil && len(d.tags) >= 1 && len(d.index) > math.MaxInt8
```

Comment:
`len(d.index) > math.MaxInt8` doesn't feel right. In particular, there's no benefit to dictionary encoding if `len(d.index) == len(d.tags)`, so I think what we want is something like `len(d.index) > len(d.tags)` or maybe `len(d.index) > len(d.tags) + nonzeroThreshold`.

Reply:
Good point